pkg/controller/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,417 @@
+	stopChan := make(chan bool, 1)
+	gvkErrdChan := make(chan schema.GroupVersionKind, 1024)
+	gvksFailingTolist := watch.NewSet()
+
+	gvksFailingToListReconciler := func(stopChan <-chan bool) {
+		for {
+			select {
+			case <-stopChan:
+				return
+			case gvk := <-gvkErrdChan:
+				gvksFailingTolist.Add(gvk)
+			}
+		}
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			close(stopChan)
+			close(gvkErrdChan)
+			return
+		case <-c.cacheManagementTicker.C:
+			c.mu.Lock()
+			c.makeUpdates(ctx)
+
+			// spin up new goroutines to relist if new gvks to relist are
+			// populated from makeUpdates.
+			if c.gvksToRelist.Size() != 0 {
+				// stop any goroutines that were relisting before
+				stopChan <- true
+
+				// also try to catch any gvks that are in the aggregator
+				// but are failing to list from a previous replay.
+				for _, gvk := range gvksFailingTolist.Items() {
+					if c.gvkAggregator.IsPresent(gvk) {
+						c.gvksToRelist.Add(gvk)
+					}
+				}
+
+				// save all gvks that need relisting
+				gvksToRelistForLoop := c.gvksToRelist.Items()
+
+				// clean state
+				gvksFailingTolist = watch.NewSet()
+				c.gvksToRelist = watch.NewSet()
+
+				stopChan = make(chan bool)
+
+				go c.replayLoop(ctx, gvksToRelistForLoop, stopChan)
+				go gvksFailingToListReconciler(stopChan)
+			}
+			c.mu.Unlock()
+		}
+	}
+}
+
+func (c *CacheManager) replayLoop(ctx context.Context, gvksToRelist []schema.GroupVersionKind, stopChan <-chan bool) {
+	for _, gvk := range gvksToRelist {
+		select {
+		case <-ctx.Done():
+			return
+		case <-stopChan:
+			return
+		default:
+			backoff := wait.Backoff{
+				Duration: time.Second,
+				Factor:   2,
+				Jitter:   0.1,
+				Steps:    3,
+			}
+
+			operation := func() (bool, error) {
+				if err := c.listAndSyncDataForGVK(ctx, gvk); err != nil {
+					return false, err
+				}
+
+				return true, nil
+			}
+
+			if err := wait.ExponentialBackoff(backoff, operation); err != nil {
+				log.Error(err, "internal: error listings gvk cache data", "gvk", gvk)
+			}
+		}
+	}
+}
+
+// listAndSyncData returns a set of gvks that were successfully listed and synced.
+func (c *CacheManager) listAndSyncData(ctx context.Context, gvks []schema.GroupVersionKind) *watch.Set {
+	gvksSuccessfullySynced := watch.NewSet()
+	for _, gvk := range gvks {
+		err := c.listAndSyncDataForGVK(ctx, gvk)
+		if err != nil {
+			log.Error(err, "internal: error syncing gvks cache data")
+			// we don't remove this gvk as we will try to re-add it later
+			// we also don't return on this error to be able to list and sync
+			// other gvks in order to protect against a bad gvk.
+		} else {
+			gvksSuccessfullySynced.Add(gvk)
+		}
+	}
+	return gvksSuccessfullySynced
+}
+
+// makeUpdates performs a conditional wipe followed by a replay if necessary as
+// given by the current spec (currentGVKsInAgg, excluderChanged) at the time of the call.
+func (c *CacheManager) makeUpdates(ctx context.Context) {
```

Comment:
naming nit: "makeUpdates" feels a bit like "doStuff".  It just means actions are taken.  Let's find a way to be more specific.